chore(vector source): Adds an RFC for vector:vector encoding/decoding and transport (#6032)
timberio/vector: rfcs/2021-01-08-5843-encdoing-decoding-format-vector.md

# RFC 5843 - 2021-01-08 - Encoding/Decoding for Vector to Vector Communication

There has been an ongoing discussion of what format Vector-to-Vector communications should take. For some examples of issues related to the subject check out #5809, #5124, or #5341. This RFC is intended to propose a format for standardization of our vector-to-vector communications.

## Scope

- This RFC should cover the encoding format we intend to leverage
- It should also cover any alterations that might need to be made to the expected data format

## Motivation

> The question posed by @binarylogic in #5843 is a good definition for the motivation of this change: "How do we transmit data from one Vector instance to another while being mindful of the public contract we create?" Clearly sending metrics from one Vector instance to another should preserve all metric data upon ingestion downstream but there are still concerns around exposing an undocumented format to users in the context of these operations. This discussion continues to come up and we really need a canonincal and definitive decision around what shape this takes so that when we expand our components our contract is explicitly codified.

Author: It's a bit odd. I think it's  _technically_ a "public" API. It's what we use to make the vector source and vector sink talk to each other and I'm assuming that users could push to or pull from this source and sink, but I don't know _why_ someone would want to. Though I'd actually probably lean on someone with a more thorough internals knowledge  than myself to fill in the gaps on that (slash correct me if I'm way off base).

> The question posed by @binarylogic in #5843 is a good definition for the motivation of this change: "How do we transmit data from one Vector instance to another while being mindful of the public contract we create?" Clearly sending metrics from one Vector instance to another should preserve all metric data upon ingestion downstream but there are still concerns around exposing an undocumented format to users in the context of these operations. This discussion continues to come up and we really need a canonincal and definitive decision around what shape this takes so that when we expand our components our contract is explicitly codified.

Reviewer: I ask because if we document the API as being internal-only then we're beholden only to ourselves for maintaining compatibility. That gives us more freedom to experiment with an approach and back off it if we need to, is why I ask.

> The question posed by @binarylogic in #5843 is a good definition for the motivation of this change: "How do we transmit data from one Vector instance to another while being mindful of the public contract we create?" Clearly sending metrics from one Vector instance to another should preserve all metric data upon ingestion downstream but there are still concerns around exposing an undocumented format to users in the context of these operations. This discussion continues to come up and we really need a canonincal and definitive decision around what shape this takes so that when we expand our components our contract is explicitly codified.

Author: Absolutely. I'm not certain if @lukesteensen has any feelings on this. I'd hate to limit our users from something that could be useful for them but I have a hard time seeing this as anything other than an internal communication API and as such I think it makes sense to document this as internal only.

> The question posed by @binarylogic in #5843 is a good definition for the motivation of this change: "How do we transmit data from one Vector instance to another while being mindful of the public contract we create?" Clearly sending metrics from one Vector instance to another should preserve all metric data upon ingestion downstream but there are still concerns around exposing an undocumented format to users in the context of these operations. This discussion continues to come up and we really need a canonincal and definitive decision around what shape this takes so that when we expand our components our contract is explicitly codified.

Reviewer 2: The intent is definitely for this to be internal-only. We don't want to introduce yet another log/metric data format into the world, but we do want something that maps pretty directly to our internal data representation and allows us the flexibility to evolve that representation as needed.

- Performance
- Compatibility with Transports
- Maintaining backcompat
- Tooling and Maintenance costs

These four concerns don't fully enumerate the problem but I think they're a really excellent distillation of what we need to keep in mind. So, to reiterate - our ideal solution solves for all four but that might just not be possible.

It might not be totally clear from the serialization benchmark examples but while protobufs are inherently slow in some respects, they're _still_ more performant than raw JSON implementations and even MessagePack or CBOR. And their maintenance cost has already been paid since we're using them today. Any schema based data format is likely going to have better perf than something schemaless. The schema itself is obviously part of the tooling overhead here but I'd argue that if we cared _more_ about tooling overhead than we did about perfomance and sustainability we'd probably be writing this project in something other than Rust.

> With regards to the tansport and the suggested path of implementing gRPC we have to keep both perf and kubernetes issues in mind. [TCP causes us some problems in our K8s integration today](https://github.com/timberio/vector/issues/2070) and unfortunately this problem has quite a bit more context to it than can sanely be shared in this RFC. Suffice to say that any choice we make here has repurcussions on our deployment architecture and ultimately gRPC with Protobufs provides us what we think are the right tradeoffs for our implementation. 

Reviewer: Woof, yes. We're not the only people that have these issues but k8s load balancing makes for a strange environment to do gRPC in. _Because_ we're not the only ones with these issues I'd expect resolution eventually -- running headless or inside a service mesh seem to be common now -- but it's for sure a foot gun for users.

Author: OH YEAH. @MOZGIII and I had discussed this at length when I originally wrote the RFC and indeed there will be some problems to solve there no matter what. We both kind of felt like aligning behind GRPC and Protobuf should give us the most likely guarantee that aspects of this specific problem soften as time goes on for that exact reason!

This brings us to the question of _why_ gRPC instead of HTTP/2 or even HTTP/3 for that matter. There are benefits in HTTP/3 but not _really_ around throughput performance as much as reliability and behavior. HTTP/3 being based on UDP means that in the case of fetching multiple objets simultaneously in the case of a dropped packet only the single interrupted stream is blocked as opposed to all streams being blocked head of line. While this might be useful behavior, the cost of writing and maintaining something in HTTP/3 will (likely) initially be much higher. Available libraries in Rust are fairly low-level and don't provide much in the way of quality abstraction for consumers, which doesn't even cover the major glaring issue that Http/3 as a protocol hasn't fully proliferated or become ubiquitious and novelty at this stage of the project is probably not what we want. That alone makes me feel like it should be avoided initially.

> So there are some tangible benefits to using gRPC instead of just HTTP/2 specifically relating to ergonomics and maintenance which ends up being the principle motivator for this decision for me. Most everything that could want to do with gRPC out of box can be achieved with HTTP/2 in hyper. However, if using gRPC we suddenly have access to [tonic](https://github.com/hyperium/tonic) which provides some truly excellent abstractions out-of-box that could pay dividends on our tooling and maintenance overhead the further we go with it, including the maintenance and overhead of protobuf generation. As a specific and shining example, should we want or need to adopt streaming requests, bi-directional stream or mutual tls authentication. `Tonic` makes this really straightforward. Once you've implemented rpc for the service you define in your .proto's the ergonomics of the library make development a breeze. Lets look at an example. First let's start with the protofbuf file specifically:

Reviewer: I have had excellent experiences with tonic, FWIW.

Author: I've only used it in toy apps and hacking projects but I also have had really positive experiences with it in those circumstances. It's clear to see even in that context how much value it provides out of box.

> So there are some tangible benefits to using gRPC instead of just HTTP/2 specifically relating to ergonomics and maintenance which ends up being the principle motivator for this decision for me. Most everything that could want to do with gRPC out of box can be achieved with HTTP/2 in hyper. However, if using gRPC we suddenly have access to [tonic](https://github.com/hyperium/tonic) which provides some truly excellent abstractions out-of-box that could pay dividends on our tooling and maintenance overhead the further we go with it, including the maintenance and overhead of protobuf generation. As a specific and shining example, should we want or need to adopt streaming requests, bi-directional stream or mutual tls authentication. `Tonic` makes this really straightforward. Once you've implemented rpc for the service you define in your .proto's the ergonomics of the library make development a breeze. Lets look at an example. First let's start with the protofbuf file specifically:

Reviewer 3: I have read the "Once you've implemented" sentence below several times, and I am completely failing to parse it. Please clarify.
```suggestion
So there are some tangible benefits to using gRPC instead of just HTTP/2 specifically relating to ergonomics and maintenance which ends up being the principle motivator for this decision for me. Most everything that we could want to do with gRPC out of box can be achieved with HTTP/2 in hyper. However, using gRPC also gives us access to [tonic](https://github.com/hyperium/tonic) which provides some truly excellent abstractions out-of-box that could pay dividends on our tooling and maintenance overhead the further we go with it, including the maintenance and overhead of protobuf generation. As a specific and shining example, should we want or need to adopt streaming requests, bi-directional stream or mutual TLS authentication. `Tonic` makes this really straightforward. Once you've implemented RPC for the service you define in your .proto's the ergonomics of the library make development a breeze.

Lets look at an example. First let's start with the protofbuf file specifically:
```

As an intentionally buried lede - I wrote this in the hopes that folks would share their opinions on the subject so you might be able to find some gaps in theorem here.

## Prior Art

Per the data format itself - I haven't found any single conglomerate conlusion about _any_ encoding format. However, existing projects using HTTP/2 and Protobufs (or gRPC over Tonic and Protobufs) are about a dime a dozen. These protocols and formats are used together in just about anything anymore.

## Drawbacks

The only drawback to adopting protobuf as the encoding format is that Protobufs can be slightly slower than other schema'd data formats and we don't shed the protobuf tooling overhead.

Drawbacks to adopting HTTP/2 are that (like literally any other transport we could pick) it alters our kubernetes deployment handling and if we don't decide to continue to maintain other transports it means needing to appropriately handle the deprecation and removal of our TCP and HTTP/1.1 implementations fairly soon afterwards.

## Alternatives

> One alternative to consider is that we just commit to providing multiple transports and stick with protobufs for encoding decoding. I think this approach should be considered thoroughly. Assuming we're using protobufs as the only encoding type, maintaining a few different transports might not be a huge maintenance overhead and it could give users some runtime flexibility that might add value.

Reviewer: If I put on my Enterprise Ops hat there's an appeal to me in a product that exposes different transports but a common serialization format between them.

> Another alternative as was pointed out by @MOZGIII the schema that we have currently is simple compared to the use cases that protobuf was designed for - so we could consider writing the whole of the serialization logic with some optimal, specially tailored process. This is likely not worth doing but, it's probably worth discussing. The downside of a hard-written solution is that we'll then maintain our own serialization library - which has a maintenance cost, potentially higher than any tooling overhead.

Reviewer: If we are not describing an internal-only API then we might consider that Vector-to-Vector is likely to be how we integrate with folk's private tooling. A custom format has less appeal here than protobuf or other open format.

## Outstanding Questions

Are there features in GRPC that we'd actually want to use?

Are there any other side benefits of using HTTP/2 or gRPC for transport?

## Plan Of Attack

Incremental steps that execute this change. Generally this is in the form of:

- [ ] Implement HTTP/1.1 with batching and v2 config (similar to the Lua transform)

> - [ ] Implementent HTTP/2

Reviewer 2: Does this need to be updated to match the proposal above? It seems like we've landed roughly on going straight to using tonic to implement this as a gRPC service.

Author: It certainly does! I'll get this updated and add a TL;DR section at the bottom that covers the plan of attack.

# RFC 5843 - 2021-01-08 - Encoding/Decoding for Vector to Vector Communication

There has been an ongoing discussion of what format Vector-to-Vector communications should take. For some examples of issues related to the subject check out #5809, #5124, or #5341. This RFC is intended to propose a format for standardization of our vector-to-vector communications.

## Scope

- This RFC should cover the encoding format we intend to leverage
- It should also cover any alterations that might need to be made to the expected data format

## Motivation

> The question posed by @binarylogic in #5843 is a good definition for the motivation of this change: "How do we transmit data from one Vector instance to another while being mindful of the public contract we create?" Clearly sending metrics from one Vector instance to another should preserve all metric data upon ingestion downstream but there are still concerns around exposing a proprietary format to users in the context of these operations. This discussion continues to come up and we really need a canonincal and definitive decision around what shape this takes so that when we expand our components our contract is explicitly codified.

Reviewer 4: ```suggestion
The question posed by @binarylogic in #5843 is a good definition for the motivation of this change: "How do we transmit data from one Vector instance to another while being mindful of the public contract we create?" Clearly sending metrics from one Vector instance to another should preserve all metric data upon ingestion downstream but there are still concerns around exposing an undocumented format to users in the context of these operations. This discussion continues to come up and we really need a canonincal and definitive decision around what shape this takes so that when we expand our components our contract is explicitly codified.
```

I might be mistaken, but I don't think the format is really "proprietary". I usually think of some sort of legal protection when that word is used.

Author: Indeed you are correct.

## Internal Proposal

From a high level my proposal is pretty simple. I'll dive into the rationale and context lower in the doc but lets just tee that up, shall we?

> I propose that if we're not searching explicitly for better perf around these messages and we're not explicitly searching for alternate tooling because we've decided that we wholly can't deal with the tooling overhead of protobufs we instead continue to use them with prost. I would also posit that if the concerns around exposing our private data format are bothersome enough, we should consider following an open spec for the data as a primary example: [opentelemetry](https://opentelemetry.io/).

Reviewer 3: I would tend to agree. With regards to opentelemetry, while it does provide an interesting target for compatibility with other systems, it does not present the same model of data as our internal data. As such using it would require either an expensive import/export procedure (relative to a protobuf codec) or reworking our internal data model to work directly with OT data.

Reviewer 5: Then I think we can safely remove OT from consideration.

> Should performance be a heavier concern for this standardization than I have assumed for the writing of this doc, my propsal would be to consider a highly performant alternative to protobufs like Flatbuffers and still consider following an open spec for the data. The `Rationale` section below will cover my thoughts on this more thoroughly.

Reviewer 5: Performance is always a top concern with Vector, but I get the sense we have a lot of internal work to do before this is the bottleneck. If that is not true, then we should seriously consider performance in the decision here. Do you have any sense of this?

cc @lukesteensen since he most likely does.

Reviewer 4: I agree with this. I think this component is performance sensitive given it should be the default for vector<->vector communication (like agent to aggregator). I'm also interested whether the protocol is a bottleneck though.

> Should performance be a heavier concern for this standardization than I have assumed for the writing of this doc, my propsal would be to consider a highly performant alternative to protobufs like Flatbuffers and still consider following an open spec for the data. The `Rationale` section below will cover my thoughts on this more thoroughly.

Reviewer 3: It is not clear if flatbuffers retain their performance advantage when the features that allow backwards compatibility are enabled. That is, while naive structs can be accessed much faster than protobufs, does that speed advantage remain if tables are used in their stead to allow for the changing of fields. I don't think we can take others' word on its performance unless they have tested with tables that require some decoding effort.

It's also worth considering that we have to consider both the encoding and decoding of the data equally, since vector bears both costs. While flatbuffers appears to be much faster to decode (subject to the above), but both prost and bincode are faster to encode, and bincode has a combined speed that tops flatbuffers overall.

All of this is to say we can't really trust the numbers below to be anything close to representative for our case.

Reviewer 6: Do I understand correctly that NoProto is essentially the idea of dumping the memory behind the data structures one's already using to the wire? In that sense, it would be very similar to FlatBuffers.

UPD: nope, it's a different thing - a format of it's own. We should consider the wild idea of just dumping the memory to the wire though - it might be doable, with some internal changes, and could be extremely efficient.

Reviewer 3: I'm not sure just duping the memory to the wire is feasible, or at least not without some rewriting. Consider the case of strings. They are stored as a structure containing a capacity and length and a *pointer* to a block of UTF-8 bytes. We have to stream both the structure and the bytes, but they are in separate locations. Should the pointer be rewritten to contain an offset into the written block where the bytes can be found or do we omit it and assume the bytes are immediately follow the string length? Do we need to store the capacity?

Even in something simple like a string, there are data layout decisions that go beyond just dumping out the memory. This also doesn't deal with cross-version compatibility (how to dump data from one version and reload with another) or more complex structures. We need a well designed/defined protocol with an efficient implementation, so we had best use something existing.

Reviewer 6: Oh yes, for sure not without rewrites. Thinking about it, it doesn't seem like it would be practical as a real-world implementation. However, dumping memory might be just an edge case to compare the perf to, as we can't be any quicker than that in theory.

## Doc-level Proposal

I'm fairly certain that this should not require a docs entry as it's largely internal to the tools. There is a possibility that we might want to document the data format we use if either to dissuade users from following it/using it (in the case that we don't adopt a standardized data format) or to point out that we've standardized on an open data spec that can be followed and used in other areas.

## Rationale

> Unfortunately in most of my research I've pretty well come to the conclusion that encoding formats as a domain are rife with technical and political failures. That is to say, in some regard or another, every format from JSON to Flatbuffers is going to force a tradeoff between correctness, performance, features or tooling and maintenance overhead. Inlined below here is a reasonable starting point on performance. It's taken from the [NoProto project](https://https://github.com/only-cliches/NoProto) since they're regularly updating their benchmarks. They've also done an excellent job of enumerating some of the aspects to consider when making a decision of this nature and their read on the space very much mirrors my own thinking in many ways.

Reviewer 4: ```suggestion
Unfortunately in most of my research I've pretty well come to the conclusion that encoding formats as a domain are rife with technical and political failures. That is to say, in some regard or another, every format from JSON to Flatbuffers is going to force a tradeoff between correctness, performance, features or tooling and maintenance overhead. Inlined below here is a reasonable starting point on performance. It's taken from the [NoProto project](https://github.com/only-cliches/NoProto) since they're regularly updating their benchmarks. They've also done an excellent job of enumerating some of the aspects to consider when making a decision of this nature and their read on the space very much mirrors my own thinking in many ways.
```

| **Runtime Libs**   |        |            |          |          |              |             |
| *NoProto*          |   1057 |       1437 |    47619 |    12195 |          208 |         166 |
| Apache Avro        |    138 |         51 |       52 |       37 |          702 |         336 |
| FlexBuffers        |    401 |        855 |    23256 |      264 |          490 |         309 |
| JSON               |    550 |        438 |      544 |      396 |          439 |         184 |
| BSON               |    115 |        103 |      109 |       80 |          414 |         216 |
| MessagePack        |    135 |        222 |      237 |      119 |          296 |         187 |
| **Compiled Libs**  |        |            |          |          |              |             |
| Flatbuffers        |   1046 |      14706 |   250000 |     1065 |          264 |         181 |
| Bincode            |   5882 |       8772 |     9524 |     4016 |          163 |         129 |
| Protobuf           |    859 |       1140 |     1163 |      480 |          154 |         141 |
| Prost              |   1225 |       1866 |     1984 |      962 |          154 |         142 |

> Performance is a single aspect of a nuanced problem for us; for more perf benchmarks I'd suggest checking out [rust-serialization-benchmarks](https://github.com/erickt/rust-serialization-benchmarks).  Some of these formats (JSON for example) still leaves us with a marginal concern around both dropped metadata as well as incidental exposure of a private data format to our users. The goal here is to hopefully solve for all of these things with whatever path is chosen.

Reviewer 4: I'm not really worried about exposing the data format to users and, as an open source project, we do that by default. We can document that it is internal and may change at any time.

While it is tempting to look at perf benchmarks and to just pick the best performing example, the practical requirements around the problem for us might give us some limitations. As was [expressed by @lukesteensen](https://discord.com/channels/742820443487993987/746070604192415834/796425792875397120) in the #development channel in discord we have three specific concerns around the choice:

> -  Our internal data model itself (inclusive of our desire to not incidentally create _another_ observability data format)

Reviewer 4: I agree that it'd be nice to leverage an existing format if it makes sense, but I don't think it is a hard requirement of internal vector <-> vector communication. For this, I think the priorities are:

* Performance
* Maintaining compatibility (e.g. if an old version of vector is sending to a newer one)
* Compatibility with a variety of transports (e.g. https://github.com/timberio/vector/issues/5124 mentions wanting to use HTTP)

Reviewer 3: Not necessarily in that order. I think backwards compatibility is, or at least will become, the top priority, followed by performance and compatibility with different transports.

> -  How we serialize that data for internal use

Reviewer 3: ITYM how we serialize the data for communication with another Vector instance, since we don't use the serialized version internally otherwise.

Reviewer 6: It is quite possible though that we'd want pick different tradeoffs for disk buffers than to the cross-instance communication. While it's an excellent high-level overview, we should probably view the encoding problem together with the expected usage patterns we expect to use it for. I.e. disk buffers can have random reads (sometimes), but flushing takes longer due to fsync (also sometimes), while the cross-instance communication, being over the network, might have different properties - like packet fragmentation (in case of non-TCP backed transport), lack of the said random reads in general, usually quicker flushes.

I guess it's absolutely nesessary to have a high-level understanding of what the different encoding formats offer, and this RFC is excellent at delivering this. My concern is that to do a leap in optimization, we might have to let go of some of the conventional abtration layers here, and look at the problem across multiple OSI layers at once.

-  The api for accessing that data model (e.g. lookups)

These three concerns don't fully enumerate the problem but I think they're a really excellent distillation of what we need to be considering. So, to reiterate - our ideal solution solves for all three but that might just not be possible. While I initially investigated this problem with really only #2 in mind, discussions with folks on the team led me to feeling like a real firecracker of a solution might be possible for both #1 and #2 out box. With the current state of lookups today, I'm less inclined to try to solve #3 immediately and I'm unconvinced that something might provide us a solution to it directly out of box. (Though, with that said, [flatbuffers](https://google.github.io/flatbuffers/) might be an interesting piece of software to take a look at in the future since it provides users the ability to access serialized data without unpacking it despite it not ultimately being my suggested approach.)

> It might not be totally clear from the serialization benchmark examples but while protobufs are inherently slow in some respects, they're _still_ more performant than raw JSON implementations and even MssagePack (which I would be willing to bet remains true for CBOR). Any schema based data format is likely going to have better perf than something schemaless. The schema itself is obviously part of the tooling overhead here but I'd argue that if we cared _more_ about tooling overhead than we did about perfomance and sustainability we'd probably be writing this project in something other than Rust.

Reviewer 3: A big :+1: on this. It's also worth noting that the tooling overhead for protobuf is largely a cost we've already paid.

Reviewer 6: I'd like to point out that the schema that we have currently in nowhere near in complexity to the use cases that protobuf was designed for - in a sense that we could possible hand-write the whole of the serialization logic with some optimal, specially tailored thing - if only we could (knew how to) do it better than protobuf and other autogenerated solutions.
The problem is it doesn't seem to be worth doing. Although... It might worth a try. At least we'll have a greater degree of control over performance.

The downside of the hard-written solution (which is really the upside of the "tooling") is that instead of tooling overhead we'll have some hand-written code - which has a maintenance cost, potentially higher than the tooling overhead, although it feels less "hacky".

Author: I'll add this to the potential alternatives.

The truth is that of the formats that have wide adoption, all of them have their pitfalls and while protobufs bring with it some real tangible maintenance burden its still some of the _best_ tools with better integrations and libraries than most of the other things out there.

As an intentionally buried lede - I wrote this in the hopes that folks would share their opinions on the subject so you might be able to find some gaps in theorem here.

## Prior Art

As mentioned above - my suggestion for the  data spec itself is that we _do_ in fact follow along with prior art. Should we follow this path, we should be giving `Open Telemetry` a thorough exploration. Per the format itself - I haven't found any single conglomerate conlusion about  _any_ encoding format. There are arguments for and against any that we could adopt.

## Drawbacks

The drawbacks to fully adopting protobufs on top of an OpenTelemetry spec are many, which makes the path no different than any other.
    - Protobufs are slower than other schema'd data formats

>     - Protobufs have an unignorable maintenance and tooling overhead

Reviewer 4: Maybe it's just me, but I haven't felt this. I also feel like our data model doesn't change very frequently such that updates to the protobuf files are rare. See https://github.com/timberio/vector/commits/master/proto/event.proto

    - Protobufs are a heavy-weight solution to what could be percieved as a lightweight problem
    - OpenTelemetry's data spec could be dramatically different than the way we structure our data today.

Without question the existing burden of protobufs continues to exist if we follow this path.

## Alternatives

> One alternative as mentioned already is flatbuffers and I didn't suggest this as a primary course of action for a few reasons. Like a lot of alternative encoding formats out there I'm honestly just not all that aware of what level of adoption it has. Also, While it has Rust support it seems rather lightweight and the level of integration available might not meet our needs. That being said, flatbuffers is really one of many many alternatives. Raw JSON with serde_bytes is another way to attack the problem. The alternatives are practically bottomless and to be perfectly honest, pretty subjective so it felt like the right approach here was to acknowledge that and submit an RFC to have a more structured conversation about those alternatives.

Reviewer 4: Should we also consider the alternative @lukesteensen suggested in https://github.com/timberio/vector/issues/5124#issuecomment-730596611 of HTTP + CBOR?

> One alternative as mentioned already is flatbuffers and I didn't suggest this as a primary course of action for a few reasons. Like a lot of alternative encoding formats out there I'm honestly just not all that aware of what level of adoption it has. Also, While it has Rust support it seems rather lightweight and the level of integration available might not meet our needs. That being said, flatbuffers is really one of many many alternatives. Raw JSON with serde_bytes is another way to attack the problem. The alternatives are practically bottomless and to be perfectly honest, pretty subjective so it felt like the right approach here was to acknowledge that and submit an RFC to have a more structured conversation about those alternatives.

Reviewer 3: I don't see flatbuffers as being substantially better than protobuf in terms of tooling cost. From my reading, they both require an external schema that needs to be compiled/translated into Rust to be useful.

## Outstanding Questions

> Should we abandon this proposal and instead do something entirely different?

Reviewer 2: I think we should expand the scope somewhat to include transport as well. I'm most interested in exploring the broader pros/cons of moving protobuf to HTTP (GRPC? Tonic?) vs doing the same with JSON. We could even consider making these "vector" sources and sinks a thin veneer over something like the Splunk HEC format if it lets us losslessy represent our internal data. Basically, we want to thoroughly address the issues with the current Vector source/sink with as little engineering effort as possible.

> Should we continue on with protobufs and solve for our data structure leakage?

Reviewer 2: It's worth being a little more specific about the "leakage" problem. In one sense, this is unavoidable since we're open source and sending this data our across the network. The things that we really want to avoid are:

1. Positioning this in any way as a new, public o11y data format
2. Coupling the serialized format so closely to our Rust code that we inadvertently change it (this is a potential downside of serde-based solutions)
3. Spending too much effort reinventing a wheel

## Scope

- This RFC should cover the encoding format we intend to leverage
- It should also cover any alterations that might need to be made to the expected data format

## Motivation

The question posed by @binarylogic in #5843 is a good definition for the motivation of this change: "How do we transmit data from one Vector instance to another while being mindful of the public contract we create?" Clearly sending metrics from one Vector instance to another should preserve all metric data upon ingestion downstream but there are still concerns around exposing an undocumented format to users in the context of these operations. This discussion continues to come up and we really need a canonincal and definitive decision around what shape this takes so that when we expand our components our contract is explicitly codified.

## Internal Proposal

From a high level my proposal is pretty simple. I'll dive into the rationale and context lower in the doc but lets just tee that up, shall we?

> I propose that if we're not explicitly searching for alternate tooling because we've decided that we wholly can't deal with the tooling overhead of protobufs, we should instead continue to use them with prost. As an added dimension to the issue I'd suggest that we support an initial Transport of HTTP/1.1 and consider a fast-follow of effort to implement HTTP/2 (or even optionally GRPC - though it bares saying I am unconvinced of the necessity of GRPC over HTTP/2 for our usecase).

Reviewer 2: Could we go into a little more detail about how we want to marry protobuf with HTTP? Right now we just have protobuf definitions for individual events, which we certainly aren't going to send one at a time each in their own request, so we will need some way to batch up those individual messages. Do you think we should define a new protobuf type for batches of events or similar?

I think the argument in favor of gRPC is less about the additional features it would provide and more about not rolling our own implementation of something that we could just take off the shelf. I'm also not convinced it's worth it, but I do think it's worth investigating to some level of depth so that we can make an informed decision. If it's as simple as using `tonic` and implementing a trait or two, that could be a compelling option over doing our own manual protobuf over HTTP (especially given features like streaming that we may want to implement in the future). Ideally these are the kinds of unknowns and future risks that we can try to suss out at the RFC phase, before we're already down the road of implementation :smile:

Reviewer 6: I agree, it is important to understand the trade-offs. I think we should investigate how do the gRPC and of the HTTP/2 compare with, say, commonly known TCP and HTTP/1. Also, HTTP/3 is a thing - we might want to consider it too.

- Performance
- Compatibility with Transports
- Maintaining backcompat
- Tooling and Maintenance costs

These four concerns don't fully enumerate the problem but I think they're a really excellent distillation of what we need to keep in mind. So, to reiterate - our ideal solution solves for all four but that might just not be possible.

It might not be totally clear from the serialization benchmark examples but while protobufs are inherently slow in some respects, they're _still_ more performant than raw JSON implementations and even MessagePack or CBOR. And their maintenance cost has already been paid since we're using them today. Any schema based data format is likely going to have better perf than something schemaless. The schema itself is obviously part of the tooling overhead here but I'd argue that if we cared _more_ about tooling overhead than we did about perfomance and sustainability we'd probably be writing this project in something other than Rust.

> With regards to the tansport and the suggested path of implementing HTTP/1.1 _before_ HTTP/2 we have to keep both perf and kubernetes issues in mind. [TCP causes us some problems in our K8s integration today](https://github.com/timberio/vector/issues/2070) and unfortunately this problem has quite a bit more context to it than can sanely be shared in this RFC. Suffice to say that any choice we make here has repurcussions on our deployment architecture and ultimately HTTP/2 with Protobufs provides us what we think are the right tradeoffs for our implementation. However, just quickly getting out a low-effort HTTP/1.1 implementation with batching solves some immediate pain for ourselves and our users around dynamic IPs and kubernetes load-balancing. I'd classify it as low-hanging fruit while a more thorough HTTP/2 and optimized Protobufs effort gets underway.

Reviewer 6: Just noting that it is highly likely that the mechanisms we'd build to properly deploy the HTTP/2 should work for other protocols as well. We should land them as reusable bits.

Another thing - we have `http` source and `http` sink, that work with HTTP/1.1 already, but are, apparently, incompatible, as the sink does batching, while the source doesn't understand it.
